### ISSUE TITLE
sig is invalid

### ISSUE THREAD

**Issue by libBletchley** [skx/pwsafe#2]
Following these instructions on debian stretch (although intended for wheezy):

https://packages.steve.org.uk/pwsafe/

I had a problem verifying the signature:

W: GPG error: http://packages.steve.org.uk/pwsafe/wheezy ./ Release: The following signatures were invalid: 36619DAAB8E831471BB1A3EFF3E8C641DC2698A1
E: The repository 'http://packages.steve.org.uk/pwsafe/wheezy ./ Release' is not signed.
E: Failed to download some files

**Comment by skx** [comment]
I've updated the signature now, which I think should fix your problem.

**NOTE**: I was using these packages as [per this blog-post](https://blog.steve.fi/so_i_have_a_wheezy_desktop.html), but these days I've migrated to [pass](https://www.passwordstore.org/) so there is zero support offered.

I should update this repository to make that clear, or delete it.

**Comment by libBletchley** [comment]
Thanks for your effort skx, but it does not seemed to have worked.  I still get this when doing aptitude update:

W: GPG error: tor://packages.steve.org.uk/pwsafe/wheezy ./ Release: The following signatures were invalid: 36619DAAB8E831471BB1A3EFF3E8C641DC2698A1
E: The repository 'tor://packages.steve.org.uk/pwsafe/wheezy ./ Release' is not signed.
E: Failed to download some files

I wonder if you updated the sig on the package, but not the sig on your repository? 

I also repeated these steps:

wget http://packages.steve.org.uk/pwsafe/apt-key.pub
apt-key add apt-key.pub

but it made no difference.

**Comment by skx** [comment]
I've only just noted your `tor:/` prefix.  I wonder if that is related?

I've downloaded both files:

    $ wget http://packages.steve.org.uk/pwsafe/wheezy/Release.gpg
    $ wget http://packages.steve.org.uk/pwsafe/wheezy/Release

Once I've done that I could verify the signature, as expected:

     $ gpg --verify Release.gpg Release
     gpg: Signature made Fri 21 Jul 2017 09:38:49 AM EEST using DSA key ID DC2698A1
     gpg: Good signature from "steve.org.uk APT key (This key is only used to sign the APT repository at http://www.steve.org.uk/apt/) <apt@steve.org.uk>"
     gpg: WARNING: This key is not certified with a trusted signature!
     gpg:          There is no indication that the signature belongs to the owner.
     Primary key fingerprint: 3661 9DAA B8E8 3147 1BB1  A3EF F3E8 C641 DC26 98A1

Can you repeat & verify that? 

* If you can verify the signature manually then something weird is going on with Tor.
* If you cannot verify the signature manually then that might indicate another local gpg-problem

**Comment by libBletchley** [comment]
I also appreciate the tip for using "pass".  The Android client for this needs Android 4.0.3, so that's a show-stopper for me.

**Comment by libBletchley** [comment]
I get this:

gpg: Signature made Fri 21 Jul 2017 08:38:49 AM CEST
gpg:                using DSA key 36619DAAB8E831471BB1A3EFF3E8C641DC2698A1
gpg:                issuer "apt@steve.org.uk"
gpg: Can't check signature: No public key

The key I have would be: http://packages.steve.org.uk/pwsafe/apt-key.pub.  Apparently that's not the correct one.

**Comment by skx** [comment]
Good edit - I was going to suggest the same thing (as you'd previously added it to the `apt`-keyring, not your personal one)

If it verifies manually then I believe that the repository is OK, and this seems like it is an issue with using tor - something I cannot help you with, I'm afraid.

**Comment by libBletchley** [comment]
I switched to http, and still get this when doing aptitude update:

W: GPG error: http://packages.steve.org.uk/pwsafe/wheezy ./ Release: The following signatures were invalid: 36619DAAB8E831471BB1A3EFF3E8C641DC2698A1
E: The repository 'http://packages.steve.org.uk/pwsafe/wheezy ./ Release' is not signed.
E: Failed to download some files

**Comment by libBletchley** [comment]
i wonder if the warning is my issue (when i verify manually):

# gpg --verify --keyring /etc/apt/trusted.gpg Release.gpg Release
gpg: Signature made Fri 21 Jul 2017 08:38:49 AM CEST
gpg:                using DSA key 36619DAAB8E831471BB1A3EFF3E8C641DC2698A1
gpg:                issuer "apt@steve.org.uk"
gpg: Good signature from "steve.org.uk APT key (This key is only used to sign the APT repository at http://www.steve.org.uk/apt/) <apt@steve.org.uk>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 3661 9DAA B8E8 3147 1BB1  A3EF F3E8 C641 DC26 98A1

**Comment by skx** [comment]
I don't believe that should matter - it just means that you have the key, but you've not explicitly trusted it.  

The signature itself is good, which is the only important thing there

**Comment by libBletchley** [comment]
I found the problem.  First of all, I would have liked to do this, but I could not because your website blocked me (perhaps a block on curl?):
```
pgpdump <(torsocks curl -s https://packages.steve.org.uk/pwsafe/wheezy/Release.gpg)
```
But if I download that file manually using a gui, and then run `pgpdump` on it, the output shows that you signed an SHA1 hash:
```
Old: Signature Packet(tag 2)(111 bytes)
        Ver 4 - new
        Sig type - Signature of a binary document(0x00).
        Pub alg - DSA Digital Signature Algorithm(pub 17)
        Hash alg - SHA1(hash 2)
        Hashed Sub: unknown(sub 33)(21 bytes)
        Hashed Sub: signature creation time(sub 2)(4 bytes)
                Time - Fri Jul 21 08:38:49 CEST 2017
        Hashed Sub: signer's User ID(sub 28)(16 bytes)
                User ID - apt@steve.org.uk
        Sub: issuer key ID(sub 16)(8 bytes)
                Key ID - 0xF3E8C641DC2698A1
        Hash left 2 bytes - a1 92 
        DSA r(160 bits) - ...
        DSA s(160 bits) - ...
                -> hash(DSA q bits)
```
SHA1 is no longer accepted by debians apt tools (which then falsely report "invalid" sig).

**Comment by libBletchley** [comment]
I noticed there is a jessie version.  I circumvented all the apt problems by taking these steps:
```
manually download https://packages.steve.org.uk/pwsafe/jessie/pwsafe_0.2.1_amd64.deb
$ dpkg -i pwsafe_0.2.1_amd64.deb
$ aptitude -f install; # to install dependencies
```
output was:
```
   The following packages have unmet dependencies:
    pwsafe : Depends: libreadline6 (>= 6.0) which is a virtual package and is not provided by any available package

             Depends: libssl1.0.0 (>= 1.0.0) which is a virtual package and is not provided by any available package

   The following actions will resolve these dependencies:

        Remove the following packages:
   1)     pwsafe [0.2.1 (now)]        

   Accept this solution? [Y/n/q/?] n

   *** No more solutions available ***
```
So we're stuffed because `libreadline6` no longer exists.

**Comment by skx** [comment]
Maybe it is time I just removed this package/repository then.

The original intention was to offer it, as it wasn't available in Debian (having been removed).  I kept it alive _here_ just for my own personal use, but now I use [pass](http://passwordstore.org/) instead..

**Comment by libBletchley** [comment]
I'm not sure if this helps with your decision, but people who migrate from your tool to the pwsafe/pwsafe project risk data corruption, which i've reported here:

https://github.com/pwsafe/pwsafe/issues/367

I'm confused about the history of these projects.  Can you say whether you forked from that project, or they forked from you?  Or was your project a fork from Bruce Schneiere's Counterpane?  I was originally using the official Debian one till it was dropped, and transitioned to yours.  But I have no idea how skx/pwsafe relates to pwsafe/pwsafe.

**Comment by skx** [comment]
My repository was created around the time this blog-post was written:

https://blog.steve.fi/so_i_have_a_wheezy_desktop.html

That matches [the commits](https://github.com/skx/pwsafe/commits/master) here which show the project built April 2013.  I didn't use a git-hub fork, I literally ran:

      apt-get source pwsafe
      git init
      git add -a 

I might have cleaned up a little, but that's basically what I did.  Looking at the repo you mention it seems that project has existed since 2001:

* https://github.com/pwsafe/pwsafe/commits/master?after=bc734dcabc32dfe4435d049f67b8238514eecedf+11510

So in short:

* Their repo is _probably_ the original/initial one.
* Mine is more recent, but is not technically a fork.
   * Just a cloned copy of the last available Debian package.

I'll reap this repository, I think.

**Comment by libBletchley** [comment]
Thanks for the history.

In principle it would make sense for your pwsafe package to be absorbed by the pwsafe/pwsafe package, which could then be an alternate CLI version sitting next to their current CLI version.  If they fix the library version problem in yours it could help troubleshoot bug https://github.com/pwsafe/pwsafe/issues/367. They could later work out which CLI version should take the lead and which should be dropped.

That's ideal, but of course finding someone interested in the effort will be the challenge.  Perhaps your repo should stick around pending that decision.
